### PR TITLE
Add support for ExperimentalAttribute

### LINKED ISSUE #6759: PublicApiAnalyzers awareness of ExperimentalAttribute annotations
The compiler has added the support for `System.Diagnostics.CodeAnalysis.ExperimentalAttribute` in https://github.com/dotnet/roslyn/pull/68702. The design is here: https://github.com/dotnet/designs/blob/main/accepted/2023/preview-apis/preview-apis.md.

It would be great to make PublicApiAnalyzers aware of the attribute. I've no idea what it would entail, starting this as a discussion.

Here are things that came to mind:

1. Annotation of the experimental API in "PublicApi.Shipped/Unshipped.txt". 
I suppose, an API could be prefixed with "[experimental]", e.g.:
    ```
    [experimental]Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer
    [experimental]Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer.UsingToStringInLoggersAnalyzer() -> void
    [experimental]override Microsoft.Extensions.ExtraAnalyzers.AsyncCallInsideUsingBlockAnalyzer.Initialize(Microsoft.CodeAnalysis.Diagnostics.AnalysisContext! context) -> void
    ```

2. An experimental API can be removed from a shipped package, so it'd be acceptable for an API recorded as an experimental to be removed from PublicApi.Shipped.txt.

3. An experimental API can be "graduated" to become a stable API, so the annotation in PublicApi.Shipped.txt and PublicApi.Unshipped.txt  can change.
    ```diff
    -  [experimental]Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer
    +  Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer
    -  [experimental]Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer.UsingToStringInLoggersAnalyzer() -> void
    +  Microsoft.Extensions.ExtraAnalyzers.UsingToStringInLoggersAnalyzer.UsingToStringInLoggersAnalyzer() -> void
    -  [experimental]override Microsoft.Extensions.ExtraAnalyzers.AsyncCallInsideUsingBlockAnalyzer.Initialize(Microsoft.CodeAnalysis.Diagnostics.AnalysisContext! context) -> void
    +  override Microsoft.Extensions.ExtraAnalyzers.AsyncCallInsideUsingBlockAnalyzer.Initialize(Microsoft.CodeAnalysis.Diagnostics.AnalysisContext! context) -> void
    ```

4. A shipped stable API can't become an experimental.
(proposed by @rafal-mz in https://github.com/dotnet/roslyn-analyzers/issues/6759#issuecomment-1632016611)

6. An experimental API should not be allowed to leak through a stable API
    ```cs
    public class StableApi
    {
       public class ExperimentalApi ProblemApi {get; set;} // Warning: experimental API leaks through stable API.

       [Experimental]
       public class ExperimentalApi OkApi {get; set;} // No warning
    }

    [Experimental]
    public class ExperimentalApi
    {
    }
    ```
    (proposed by @rafal-mz in https://github.com/dotnet/roslyn-analyzers/issues/6759#issuecomment-1632016611)

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #6759

Experimental items are shown in API files with a `[diagnosticId]` prefix, where `diagnosticId` is the ID provided to `ExperimentalAttribute`.

### CODE REVIEW

**Review by mavasani — COMMENTED** [review]
LGTM, but I’d prefer @jcouv to review it and merge once we have his approval.

**Comment by jcouv** [comment]
Looking

File: src/PublicApiAnalyzers/UnitTests/DeclarePublicAPIAnalyzerTestsBase.cs
Comment (jcouv) [inline src/PublicApiAnalyzers/UnitTests/DeclarePublicAPIAnalyzerTestsBase.cs:3246]: Consider also testing removal of Experimental attribute (shipped file contains the experiment, but API no longer marked as experimental) and change of Experimental attribute (however I'm not sure it's a realistic/common scenario)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1) with some test suggestions to consider

File: src/PublicApiAnalyzers/UnitTests/DeclarePublicAPIAnalyzerTestsBase.cs
Comment (sharwell) [reply src/PublicApiAnalyzers/UnitTests/DeclarePublicAPIAnalyzerTestsBase.cs:3246]: Currently this will be treated as simply removing an old API and adding a new one. We can work to improve this experience in the future based on how things go with Roslyn experiments.

**Comment by sharwell** [comment]
@jcouv I'm going to submit new tests based on how this first round goes when I submit a PR with an experiment to Roslyn.

**Pull request merged**
